src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -32,7 +56,91 @@ private CompletableFuture<Result<Map<RequestType, Set<String>>>> getAllowedServi
+
+    log.debug("lookupRequestPolicy:: parameters: request={}, user={}", request, user);
+
+    if (user == null) {
+      log.error("lookupRequestPolicy:: user is null");
+    }
+
+    return itemRepository.fetchById(request.getItemId())
+      .thenCompose(r -> r.after(item -> requestPolicyRepository.lookupRequestPolicy(item, user)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> extractAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("extractAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    var allowedServicePointsInPolicy = requestPolicy.getAllowedServicePoints();
+    if (allowedServicePointsInPolicy.isEmpty()) {
+      log.info("extractAllowedServicePoints:: allowedServicePointsInPolicy is empty");
+
+      return fetchAllowedServicePoints(requestPolicy);
+    }
+
+    Set<String> allowedServicePointsIdsSet = allowedServicePointsInPolicy.values().stream()
+      .flatMap(Collection::stream)
+      .map(UUID::toString)
+      .collect(Collectors.toSet());
+
+    return servicePointRepository.filterIdsByServicePointsAndPickupLocationExistence(
+      allowedServicePointsIdsSet)
+        .thenCompose(r -> r.after(filteredSpIds -> includeOnlyExistingServicePoints(
+          allowedServicePointsInPolicy, filteredSpIds, requestPolicy)));
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> fetchAllowedServicePoints(
+    RequestPolicy requestPolicy) {
+
+    log.debug("fetchAllowedServicePoints:: parameters: requestPolicy={}", requestPolicy);
+
+    List<RequestType> allowedTypes = Arrays.stream(RequestType.values())
+      .filter(requestPolicy::allowsType)
+      .toList();
+    if (!allowedTypes.isEmpty()) {
+      log.info("fetchAllowedServicePoints:: allowedTypes={}", allowedTypes);
+      return servicePointRepository.fetchPickupLocationServicePointsIds()
+        .thenCompose(r -> r.after(servicePointIds -> {
+          Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
+          allowedTypes.forEach(requestType -> allowedServicePoints.put(requestType, servicePointIds));
+
+          log.info("fetchAllowedServicePoints:: result: {}", () -> asJson(allowedServicePoints));
+
+          return ofAsync(allowedServicePoints);
+        }));
+    } else {
+      return ofAsync(new EnumMap<>(RequestType.class));
+    }
+  }
+
+  private CompletableFuture<Result<Map<RequestType, Set<String>>>> includeOnlyExistingServicePoints(
+    Map<RequestType, Set<UUID>> allowedServicePointsInPolicy, Set<String> relevantSpIds,
+    RequestPolicy requestPolicy) {
+
+    log.debug("includeOnlyExistingServicePoints:: parameters: allowedServicePointsInPolicy={}, " +
+      "relevantSpIds={}, requestPolicy={}", () -> asJson(allowedServicePointsInPolicy),
+      () -> asJson(relevantSpIds), () -> asJson(requestPolicy));
+
+    final Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
```

Comment:
Instead building a new map with relevant IDs, can we just remove irrelevant IDs from `allowedServicePointsInPolicy`?

Reply:
we don't only remove irrelevant IDs, we also convert Set<UUID> to Set<String>